earth2studio/models/px/dlesym.py

```diff
@@ -0,0 +1,1077 @@
+        slices = ocean_coupling.chunk(len(self.ocean_input_times), dim=2)
+        ocean_coupling = torch.concat(
+            [s.mean(dim=2, keepdim=True) for s in slices], dim=3
+        )
+        ocean_coupling = ocean_coupling.permute(2, 0, 3, 1, 4, 5)
+        return ocean_coupling
+
+    def _normalize_input(self, x: torch.Tensor) -> torch.Tensor:
+        """Normalize input data"""
+
+        return (x - self.center) / self.scale
+
+    def _denormalize_output(self, x: torch.Tensor) -> torch.Tensor:
+        """Denormalize output data"""
+
+        return x * self.scale + self.center
+
+    def retrieve_valid_ocean_outputs(
+        self, x: torch.Tensor, coords: CoordSystem
+    ) -> torch.Tensor:
+        """Retrieve the valid ocean model outputs from an output data tensor.
+        Because we use a dense grid of output times for the coupled model, some of the output times
+        for the ocean model may not be valid because it takes a coarser time-step.
+        This function will retrieve the valid outputs and return them in a tensor of shape (batch, time, lead_time, variable, face, height, width).
+
+        Parameters
+        ----------
+        x : torch.Tensor
+            Output data tensor
+        coords : CoordSystem
+            Input coordinates
+
+        Returns
+        -------
+        torch.Tensor
+            Ocean outputs
+        CoordSystem
+            Output coordinates
+        """
+
+        out_coords = coords.copy()
+        out_coords["variable"] = np.array(self.ocean_variables)
+        out_coords["lead_time"] = np.array(
+            [t for t in coords["lead_time"] if t % self.ocean_output_times[0] == 0]
+        )
+
+        ocean_lead_time_idx = [
+            list(coords["lead_time"]).index(t) for t in out_coords["lead_time"]
+        ]
+        ocean_outputs = x[:, :, ocean_lead_time_idx, ...]
+        ocean_var_idx = [
+            list(coords["variable"]).index(var) for var in self.ocean_variables
+        ]
+        ocean_outputs = ocean_outputs.index_select(
+            dim=3, index=torch.tensor(ocean_var_idx, device=x.device)
+        )
+        return ocean_outputs, out_coords
+
+    def retrieve_valid_atmos_outputs(
+        self, x: torch.Tensor, coords: CoordSystem
+    ) -> torch.Tensor:
+        """Retrieve the valid atmospheric model outputs from an output data tensor.
+        This function will retrieve the valid outputs and return them in a tensor of shape (batch, time, lead_time, variable, face, height, width).
+
+        Parameters
+        ----------
+        x : torch.Tensor
+            Output data tensor
+        coords : CoordSystem
+            Input coordinates
+
+        Returns
+        -------
+        torch.Tensor
+            Atmos outputs
+        CoordSystem
+            Output coordinates
+        """
+
+        out_coords = coords.copy()
+        out_coords["variable"] = np.array(self.atmos_variables)
+
+        atmos_var_idx = [
```

Comment:
I see this index search done a lot. Is this needed, or can these be hard coded once? Maybe theres a slice operation that can be used instead? Curious of how important it is to have this flexible and calced every forward pass?

Reply:
The variable indices can definitely be hard-coded in `__init__` and re-used. I think it should also be doable for the lead time indices, though it will add some clutter to `__init__` as the indexing needs to be done for the input states, coupling inputs, and output states for each of the atmos/ocean models. Probably worth it though for speed (and decluttering the other routines)

Reply:
What ever is easier / clearer to read. Up to you!